### ISSUE TITLE
Visibility of Sidecar, DestinationRule, ServiceEntry and VirtualServices changed on Istio 1.5

### ISSUE THREAD

**Issue by chadlwilson** [istio/istio#24251]
### Bug description

During upgrade from Istio `1.4.9` to Istio `1.5.4` we experienced a regression caused by apparent changes in visibility of hosts configured via Istio networking resources across namespaces.

The intention of our configuration is to control certain mesh-external `SeviceEntry`s and associated configuration to only be bootstrapped/sent to certain proxies/apps.

Instead, after upgrade we saw `DestinationRule` and `VirtualService` configuration linked to un-permissioned `ServiceEntry`s being bootstrapped to ALL proxies, instead of those that should be able to see them. 

This caused most of our proxies to **fail to become ready**; because we have `DestinationRule` configuration intended only for certain proxied apps, which requires certain mounted mTLS certs via `sidecar.istio.io` annotations required for TLS-origination on some external services. Without those, the CDS updates cannot be acked. 

For other Istio users this change may just go unnoticed, and lead to unnecessary configuration/Pilot updates having to be pushed to proxies which they should not.

**Working `1.4.9` resource structure** (failed on `1.5.4`)
- `ns1` **Namespace** - istio-injected
- `ns-api` **Namespace** Dummy namespace for external `ServiceEntry`s. No pods here.
- `istio-system` has default **Sidecar** rule
    - apps can only see configuration in same namespace + `istio-system`
- `ns1` **Sidecar** rule for app `app1` 
    - makes host `ns-api/ext-service.domain.com` from `serviceentry1.ns-api` **ServiceEntry** visible to `app1` by `workloadSelector`
- `ns1` **DestinationRule** configuration linked to host `ext-service.domain.com`
    - configures mTLS and `outlierDetection` for host
- `ns1` **VirtualService** configuration linked to host `ext-service.domain.com`
    - configures port 80 -> port 443 for host
- `ns-api` **ServiceEntry** configuration for host `ext-service.domain.com`
    - standard `resolution: DNS` `MESH_EXTERNAL` entries

Envoy cluster config dumps showed the cluster entries for `ext-service.domain.com` being sent to **every** proxy in the mesh, despite the `Sidecar` scoping noted above.

The workaround (from trial and error) that works on `1.5.4` was to move the `VirtualService` entries (but NOT the `DestinationRule` to namespace `ns-api` alongside the ServiceEntry)

**Working `1.5.4` resource structure**
- `ns1` **Namespace** - istio-injected
- `ns-api` **Namespace** Dummy namespace for external `ServiceEntry`s. No pods here.
- `istio-system` default **Sidecar** rule
    -services can only see things in same namespace + `istio-system`
- `ns1` **Sidecar** rule to make host `ext-service.domain.com` from `serviceentry1.ns-api` **ServiceEntry** visible
- `ns1` **DestinationRule** configuration linked to host `ext-service.domain.com`
- `ns-api` **VirtualService** configuration linked to host `ext-service.domain.com` **<-- CHANGED**
- `ns-api` **ServiceEntry** configuration for host `ext-service.domain.com`

**Other odd observation**

If we moved **both** the **DestinationRule** and the **VirtualService** to `ns-api` namespace it caused our `istio-system`-based ingress gateways to fail to start, as they seemed to then get the `DestinationRule` config bootstrapped to them. This was very strange to us; never worked out why this was.

### Affected product area (please put an X in all that apply)

[X] Networking

### Expected behavior

Although the workaround of moving the `VirtualService` to the same namespace as the `ServiceEntry` also worked on Istio `1.4.7`, so was forward compatible - we were not expecting this change in behaviour, and could not see anything in the changelogs which seemed to be related to changes in the way the `Sidecar` API determined which proxies in which namespaces would be able to see various configuration.

### Version 

Istio `1.5.4`
Kubernetes `1.15`
Helm `3.1.2`

### How was Istio installed?

Helm installed, Helm-upgraded to `1.5.4` using the official chart. Thus we are still using separate `pilot`, `citadel`, `galley` etc. Upgrade process is summarised [in this gist](https://gist.github.com/chadlwilson/c4653f6a5a1da43ccacf46589d7bf482).

Relevant settings
- `REGISTRY_ONLY` enabled (so `ServiceEntry`s should control ALL external egress)

### Environment where bug was observed (cloud vendor, OS, etc)

AWS EKS

**Comment by howardjohn** [comment]
Thanks for the details, I will take a closer look at this. To clarify, with 1.5 and the original config, some other random namespace gets the ext-service config if the virtual service is in ns1? Or is it other workloads in ns1 get it unexpectedly?

**Comment by chadlwilson** [comment]
@howardjohn The second of your alternatives - all the other workloads in `ns1` seemed to get cluster config for this external service host unexpectedly. They could see configuration implied by both the `VirtualService` and the `DestinationRule`.

What it looks like is that the "visibility" of a given host/cluster config in `1.5` is also being driven by the `VirtualService` rather than solel;y the `ServiceEntry` (or possibly it's an "either/or" situation - if a Sidecar can see config related to a host defined either by a `VirtualService` or a `ServiceEntry` it gets a cluster config sent to it).

It's entirely possible that we were relying on an odd configuration earlier (which might even be considered a bug...) however. If that's the case, all good - but at least it's here for posterity :-)

Also worth noting if it matters, that since this was done via an in-place Helm upgrade **we are not using istiod**. Our upgrade process is summarised [here](https://gist.github.com/chadlwilson/c4653f6a5a1da43ccacf46589d7bf482).

**Comment by howardjohn** [comment]
Ah okay I see now, so likely the virtual service is pulling in all the config. Thanks for the clarification!

**Comment by howardjohn** [comment]
Looks pretty clearly a result of https://github.com/istio/istio/pull/20408.

Based on that I would say the current behavior is the desired behavior, and this was an undocumented breaking change (mostly because we didn't consider the implication for configs like this).

**Comment by chadlwilson** [comment]
Thanks for digging that up @howardjohn - that does explain it.

Is it possible for us to add something into the release/upgrade nodes for `1.5` to warn people to prepare for this visibility change?

I assume this also explains the weird behaviour of our ingress gateways also getting this config; but I don't quite understand why if I moved the `DestionationRule` into the dummy `ns-api` namespace that the ingress-gateways started seeing the tls config that's part of the DRs too - even though they don't see the `DestinationRule` config if it's left in the main `ns` namespace. I assume it's something to do with having our ingress gateways in `istio-system` meaning they generally "see everything" - but took quite a bit of trial and error to work out I needed to leave the `DestinationRule`s where they were to avoid breaking the gateways.
